### PR TITLE
Optimize memory footprint

### PR DESCRIPTION
**Pull request by aslesarenko**
This PR implements simple refactoring to improve performance and memory footprint.

### CODE REVIEW

File: core/src/main/scala/scalan/Entities.scala [hunk 1]
```diff
@@ -28,7 +28,7 @@ trait Entities extends TypeDescs { self: Scalan =>
       case _ => false
     })
 
-    override def hashCode = Objects.hash(getClass, typeArgsDescs)
+    override def hashCode = getClass.hashCode() * 31 + typeArgsDescs.hashCode()
```
Comment (kushti) [inline core/src/main/scala/scalan/Entities.scala:31]: Remove imports not used anymore, provide ScalaDoc for the trait

File: library-impl/src/main/scala/special/collection/CollsOverArrays.scala [hunk 2]
```diff
@@ -305,7 +305,7 @@ class CollOverArrayBuilder extends CollBuilder {
       val rs = emptyColl(pt.tSnd)
       asColl[T](pairColl(ls, rs))
     case _ =>
-      new CollOverArray[T](Array[T]())
+      new CollOverArray[T](cT.emptyArray)
```
Comment (kushti) [inline library-impl/src/main/scala/special/collection/CollsOverArrays.scala:308]: ScalaDoc for the method (at least) missed

File: sigmastate/src/main/scala/org/ergoplatform/ErgoLikeContext.scala [hunk 3]
```diff
@@ -194,58 +194,58 @@ object ErgoLikeContext {
...
-  def opType = SFunc(SContext, SBox)
+  override val opType = SFunc(SContext, SBox)
 }
 
 case object Context extends NotReadyValue[SContext.type] with ValueCompanion {
   override def companion = this
   override def opCode: OpCode = OpCodes.ContextCode
   override def tpe: SContext.type = SContext
-  override def opType: SFunc = SFunc(SUnit, SContext)
+  override val opType: SFunc = SFunc(SUnit, SContext)
```
Comment (kushti) [inline sigmastate/src/main/scala/org/ergoplatform/ErgoLikeContext.scala:243]: Add ScalaDoc for the class.

File: sigmastate/src/main/scala/org/ergoplatform/ErgoLikeContext.scala [hunk 4]
```diff
@@ -194,58 +194,58 @@ object ErgoLikeContext {
...
 }
 
 case object Global extends NotReadyValue[SGlobal.type] with ValueCompanion {
   override def companion = this
   override def opCode: OpCode = OpCodes.GlobalCode
   override def tpe: SGlobal.type = SGlobal
-  override def opType: SFunc = SFunc(SUnit, SGlobal)
+  override val opType: SFunc = SFunc(SUnit, SGlobal)
```
Comment (kushti) [inline sigmastate/src/main/scala/org/ergoplatform/ErgoLikeContext.scala:250]: Add ScalaDoc for the class.

File: sigmastate/src/main/scala/sigmastate/eval/CostingDataContext.scala [hunk 5]
```diff
@@ -461,6 +462,29 @@ object CHeader {
   * @see [[CostModel]] for detailed descriptions
   */
 class CCostModel extends CostModel {
+  import CCostModel._
+
+  def AccessBox: Int = AccessBoxCost
```
Comment (kushti) [inline sigmastate/src/main/scala/sigmastate/eval/CostingDataContext.scala:467]: ScalaDoc missed here and further

File: sigmastate/src/main/scala/sigmastate/eval/RuntimeCosting.scala [hunk 6]
```diff
@@ -1687,7 +1687,7 @@ trait RuntimeCosting extends CostingRules { IR: IRContext =>
         def tC = evalNode(ctx, env, t)
         def eC = evalNode(ctx, env, e)
         val resV = IF (cC.value) THEN tC.value ELSE eC.value
-        val resCost = opCost(resV, Array(cC.cost, tC.cost, eC.cost), costOf("If", SFunc(Vector(SBoolean, If.tT, If.tT), If.tT)))
+        val resCost = opCost(resV, Array(cC.cost, tC.cost, eC.cost), costOf("If", If.GenericOpType))
```
Comment (kushti) [inline sigmastate/src/main/scala/sigmastate/eval/RuntimeCosting.scala:1690]: add ScalaDoc for evalNode, better comment  branches inside as well

File: sigmastate/src/main/scala/sigmastate/eval/TreeBuilding.scala [hunk 7]
```diff
@@ -134,7 +134,7 @@ trait TreeBuilding extends RuntimeCosting { IR: IRContext =>
         val varId = defId + 1       // arguments are treated as ValDefs and occupy id space
         val env1 = env + (x -> (varId, elemToSType(x.elem)))
         val block = processAstGraph(ctx, mainG, env1, lam, varId + 1, constantsProcessing)
-        val rhs = mkFuncValue(Vector((varId, elemToSType(x.elem))), block)
+        val rhs = mkFuncValue(Array((varId, elemToSType(x.elem))), block)
```
Comment (kushti) [inline sigmastate/src/main/scala/sigmastate/eval/TreeBuilding.scala:137]: add ScalaDoc to the method

File: sigmastate/src/main/scala/sigmastate/trees.scala [hunk 8]
```diff
@@ -198,24 +201,30 @@ object SigmaAnd extends SigmaTransformerCompanion {
...
 }
 
 object SigmaOr extends SigmaTransformerCompanion {
+  val OpType = SFunc(SCollection.SSigmaPropArray, SSigmaProp)
   override def opCode: OpCode = OpCodes.SigmaOrCode
   override def argInfos: Seq[ArgInfo] = SigmaOrInfo.argInfos
   def apply(head: SigmaPropValue, tail: SigmaPropValue*): SigmaOr = SigmaOr(head +: tail)
 }
 
+/** Base trait for companions. */
```
Comment (kushti) [inline sigmastate/src/main/scala/sigmastate/trees.scala:215]: This comment is not saying really anything.

File: sigmastate/src/main/scala/sigmastate/types.scala [hunk 9]
```diff
@@ -118,6 +120,28 @@ object SType {
   }
   implicit val ErgoLikeContextRType: RType[ErgoLikeContext] = RType.fromClassTag(classTag[ErgoLikeContext])
 
+  /** Named type variables and parameters used in generic types and method signatures. */
+  val tT = STypeVar("T")
```
Comment (kushti) [inline sigmastate/src/main/scala/sigmastate/types.scala:124]: what does all these letters mean?

**Review by kushti — CHANGES_REQUESTED** [review]

File: core/src/main/scala/scalan/Entities.scala (hunk 1 above)
Comment (aslesarenko) [reply core/src/main/scala/scalan/Entities.scala:31]: done

File: library-impl/src/main/scala/special/collection/CollsOverArrays.scala (hunk 2 above)
Comment (aslesarenko) [reply library-impl/src/main/scala/special/collection/CollsOverArrays.scala:308]: This is override, which means the ScalaDoc is in the super class.

File: sigmastate/src/main/scala/org/ergoplatform/ErgoLikeContext.scala (hunk 3 above)
Comment (aslesarenko) [reply sigmastate/src/main/scala/org/ergoplatform/ErgoLikeContext.scala:243]: done

File: sigmastate/src/main/scala/org/ergoplatform/ErgoLikeContext.scala (hunk 4 above)
Comment (aslesarenko) [reply sigmastate/src/main/scala/org/ergoplatform/ErgoLikeContext.scala:250]: done

File: sigmastate/src/main/scala/sigmastate/eval/CostingDataContext.scala (hunk 5 above)
Comment (aslesarenko) [reply sigmastate/src/main/scala/sigmastate/eval/CostingDataContext.scala:467]: added ScalaDocs in the CostModel trait and marked methods as `override`.
Some methods are not used, so marked them as `TODO refactor: ...`

File: sigmastate/src/main/scala/sigmastate/eval/RuntimeCosting.scala (hunk 6 above)
Comment (aslesarenko) [reply sigmastate/src/main/scala/sigmastate/eval/RuntimeCosting.scala:1690]: Added ScalaDoc for the method. Note, this entire file will be removed in v5.0.

File: sigmastate/src/main/scala/sigmastate/eval/TreeBuilding.scala (hunk 7 above)
Comment (aslesarenko) [reply sigmastate/src/main/scala/sigmastate/eval/TreeBuilding.scala:137]: done

File: sigmastate/src/main/scala/sigmastate/types.scala (hunk 9 above)
Comment (aslesarenko) [reply sigmastate/src/main/scala/sigmastate/types.scala:137]: Added description.

File: sigmastate/src/main/scala/sigmastate/trees.scala (hunk 8 above)
Comment (aslesarenko) [reply sigmastate/src/main/scala/sigmastate/trees.scala:215]: Well it is better then nothing. Added more details.

**Review by kushti — APPROVED** [review]

**Pull request merged**